tests/raster/test_geotiff.py

```diff
@@ -0,0 +1,307 @@
+import os
+import shutil
+import unittest
+import numpy as np
+import pandas as pd
+import xarray as xr
+from tempfile import mkdtemp
+
+from geospade.crs import SpatialRef
+from geospade.raster import Tile
+from geospade.raster import MosaicGeometry
+
+from veranda.raster.native.geotiff import GeoTiffFile
+from veranda.raster.mosaic.geotiff import GeoTiffReader, GeoTiffWriter
+
+
+class GeoTiffFileTest(unittest.TestCase):
+
+    def setUp(self):
+        """
+        Set up dummy mosaic set.
+        """
+        self.filepath = os.path.join(mkdtemp(), 'test.tif')
+
+    def test_read_write_1_band(self):
+        """
+        Test a write and read of a single band mosaic.
+        """
+        data = np.ones((1, 100, 100), dtype=np.float32)
+
+        with GeoTiffFile(self.filepath, mode='w') as src:
+            src.write(data)
+
+        with GeoTiffFile(self.filepath) as src:
+            ds = src.read()
+
+        np.testing.assert_array_equal(ds[1], data[0, :, :])
+
+    def test_read_write_multi_band(self):
+        """
+        Test write and read of a multi band mosaic.
+        """
+        data = np.ones((5, 100, 100), dtype=np.float32)
+
+        with GeoTiffFile(self.filepath, mode='w', n_bands=5) as src:
+            src.write(data)
+
+        with GeoTiffFile(self.filepath) as src:
+            for band in np.arange(data.shape[0]):
+                ds = src.read(bands=band + 1)
+                np.testing.assert_array_equal(
+                    ds[band + 1], data[band, :, :])
+
+    def test_decoding_multi_band(self):
+        """
+        Test decoding of multi band mosaic.
+
+        """
+        data = np.ones((5, 100, 100), dtype=np.float32)
+        scale_factors = {1: 1, 2: 2, 3: 1, 4: 1, 5: 3}
+        offsets = {2: 3}
+        with GeoTiffFile(self.filepath, mode='w', n_bands=5,
+                         scale_factors=scale_factors, offsets=offsets) as src:
+            src.write(data)
+
+        with GeoTiffFile(self.filepath, auto_decode=False) as src:
+            ds = src.read(bands=1)
+            np.testing.assert_array_equal(ds[1], data[0, :, :])
+            ds = src.read(bands=2)
+            np.testing.assert_array_equal(ds[2], data[1, :, :])
+            ds = src.read(bands=5)
+            np.testing.assert_array_equal(ds[5], data[4, :, :])
+
+        data[1, :, :] = data[1, :, :] * 2 + 3
+        data[4, :, :] = data[4, :, :] * 3
+
+        with GeoTiffFile(self.filepath, auto_decode=True) as src:
+            ds = src.read(bands=1)
+            np.testing.assert_array_equal(ds[1], data[0, :, :])
+            ds = src.read(bands=2)
+            np.testing.assert_array_equal(ds[2], data[1, :, :])
+            ds = src.read(bands=5)
+            np.testing.assert_array_equal(ds[5], data[4, :, :])
```

Comment:
should be 2 separate tests `test_auto_decode_multi_band` and `test_ignore_decoding_multiband`

Reply:
Done.